tests/sentry/auth/test_helper.py

```diff
@@ -77,25 +120,147 @@ def test_associated_existing_member_invite_request(self):
             {"memberId": member.id, "token": member.token, "url": ""}
         )
 
-        auth_identity = handle_new_user(provider, self.organization, request, identity)
+        auth_identity = self._handle_new_user()
 
         assigned_member = OrganizationMember.objects.get(
             organization=self.organization, user=auth_identity.user
         )
 
         assert assigned_member.id == member.id
+
+
+class HandleExistingIdentityTest(AuthIdentityHandlerTest):
+    @mock.patch("sentry.auth.helper.auth")
+    def test_simple(self, mock_auth):
+        mock_auth.get_login_redirect.return_value = "test_login_url"
+        user, auth_identity = self.set_up_user_identity()
+
+        redirect = handle_existing_identity(
+            self.auth_provider,
+            self.provider_obj,
+            self.organization,
+            self.request,
+            self.state,
+            auth_identity,
+            self.identity,
+        )
+
+        assert redirect.url == mock_auth.get_login_redirect.return_value
+        assert mock_auth.get_login_redirect.called_with(self.request)
+
+        persisted_identity = AuthIdentity.objects.get(ident=auth_identity.ident)
+        assert persisted_identity.data == self.identity["data"]
+
+        persisted_om = OrganizationMember.objects.get(user=user, organization=self.organization)
+        assert getattr(persisted_om.flags, "sso:linked")
+        assert not getattr(persisted_om.flags, "sso:invalid")
+
+        login_request, login_user = mock_auth.login.call_args.args
+        assert login_request == self.request
+        assert login_user == user
+
+
+class HandleAttachIdentityTest(AuthIdentityHandlerTest):
+    def _handle_attach_identity(self):
+        return handle_attach_identity(
+            self.auth_provider, self.request, self.organization, self.provider_obj, self.identity
+        )
+
+    @mock.patch("sentry.auth.helper.messages")
+    def test_new_identity(self, mock_messages):
+        self.set_up_user()
+
+        auth_identity = self._handle_attach_identity()
+        assert auth_identity.ident == self.identity["id"]
+        assert auth_identity.data == self.identity["data"]
+
+        assert AuditLogEntry.objects.filter(
+            organization=self.organization,
+            target_object=auth_identity.id,
+            event=AuditLogEntryEvent.SSO_IDENTITY_LINK,
+            data=auth_identity.get_audit_log_data(),
+        ).exists()
+
+        assert mock_messages.add_message.called_with(
+            self.request, messages.SUCCESS, OK_LINK_IDENTITY
+        )
+
+    @mock.patch("sentry.auth.helper.messages")
+    def test_existing_identity(self, mock_messages):
+        user, existing_identity = self.set_up_user_identity()
+
+        returned_identity = self._handle_attach_identity()
+        assert returned_identity == existing_identity
+        assert not mock_messages.add_message.called
+
+    def test_wipe_other_identity(self):
```

Comment:
what about the other case of this? there is no associated identity in DB for the request user, but the request contains an id that matches an existing one

Reply:
Good question. The only behavioral difference is that, in the case you're describing not currently covered, the query
```python
AuthIdentity.objects.exclude(id=auth_identity.id).filter(
    auth_provider=self.auth_provider, user=self.user
)
```
returns an empty result set and therefore deletes nothing. So the same test but with a slightly smaller setup and one fewer assertion.

Still worth splitting out into two cases, though! Will do.